mechanicalsoup/stateful_browser.py

```diff
@@ -216,6 +221,19 @@ def submit_selected(self, btnName=None, *args, **kwargs):
         to :func:`Form.choose_submit` on the current form to choose between
         them. All other arguments are forwarded to :func:`Browser.submit`.
         """
+        # Temporarily allow calling the old inherited Browser.submit directly
+        # in case we can detect with certainty that the call is an old style.
+        # Note: Browser.submit also accepts a bs4.element.Tag with name="form",
+        # but we cannot assume this is an old-style call since there could be
+        # a submit button with name="form".
+        if isinstance(btnName, Form):
+            warnings.warn("This usage of StatefulBrowser.submit is deprecated."
+                          " Please see the documentation for this function to "
+                          "upgrade to the new interface.",
+                          DeprecationWarning)
```

Comment:
Unfortunately, forbidding this is not only a backward-compatibility problem, but also a violation of Liskov's Substitution Principle. The base `.submit()` accepts a `btnName` instance of `Form`, so all derived methods should also accept at least that (and possibly more), so that anything doable with a `Browser` is also doable with a `StatefulBrowser`.

Also, having the same positional argument be `btnName` in one class and `form` in another is very misleading. LGTM is right to complain here IMHO.

Reply:
Before I invest too much more time into this, I just wanted to check if you are generally opposed to this PR, or if you can imagine a modified PR that you would consider accepting. Even better if you already have an idea in mind! ;)

My preference is always to make `StatefulBrowser` the best class it can be. The `Browser` class sometimes gets in the way of that, such as in this case. I believe `StatefulBrowser` would benefit from a `submit` method, so if we can't override `Browser.submit` in the proposed way, the first two alternatives that come to mind (setting aside the issue of backwards compatibility for a moment) are renaming `Browser.submit` to either a) `Browser._submit` or b) `Browser.submit_form`.

Reply:
If I were to redo this from scratch, I'd consider using composition instead of inheritance, i.e. consider that a `StatefulBrowser` *contains* a `Browser`, not that it *is* a `Browser`. It'd be hard to change that without breaking backward compatibility, though.

Renaming `Browser.submit` is also problematic with respect to backward compatibility. Perhaps not many people call `submit` on a `StatefulBrowser`, but everyone who wrote code before `StatefulBrowser` was added does call `submit` on a `Browser`.

There's another option: add warnings for suspicious cases, and allow the user to disable the warnings as needed. We could warn on calls to `StatefulBrowser.submit`, or if we want to go further, on creation of a `Browser` that isn't a `StatefulBrowser`. Probably not as a deprecation warning like "what you're doing won't work anymore soon", but just like "uh, are you sure you want that?".

Reply:
A warning for suspicious use is a great idea, but I'm not sure that it addresses the main issues here (and might be just as much of an annoyance as a modified interface to some users). Out of curiosity, though, did you have a design in mind for warning toggling?

Okay, two more possibilities:
1. Simply add a comment to the `Browser.submit` docstring saying something like:
> If you are calling this method from a `StatefulBrowser` instance, consider using `StatefulBrowser.submit_selected` instead.

2. Override `Browser.submit` in a Liskov-compliant way, e.g.
```python
def submit(self, form=None, url=None, **kwargs, btnName=None):
```
where `form` and `url` default to the current form/url if `None`. It's a bit of a clunky interface, but might be the best we can do if we want to make overriding work within our constraints. Or even just
```python
def submit(self, form=None, url=None, **kwargs):
```
to avoid complications with `btnName`.

Reply:
@moy I'd like to get version 0.11 released as soon as possible, but I want to address this issue in some way before I do, even if it's just a minor addition to the documentation.

How about I make the documentation change proposed above for 0.11, and then we can reassess if we want to make any code changes for version 1.0 later?

Reply:
The doc change is obviously a good step forward.

Your option 2. is probably doable too. I won't have time to look at this in details soon, but if you're confident enough I trust you to do the right thing.

Reply:
Thanks for the feedback! For the 0.11 release I'll just change the docstring, unless you recommend otherwise.

As I think more about it, overriding `submit` seems increasingly important. One of the biggest issues (that I didn't appreciate until just now) is that if you call `submit` on a StatefulBrowser, the browser state becomes stale. I would argue that this is a pretty dangerous inconsistency for a method that looks like the most obvious way to submit a form.